test/Microsoft.ML.Core.Tests/UnitTests/TestVBuffer.cs

```diff
@@ -21,5 +31,1174 @@ public void TestApplyAt()
+            for (int i = 0; i < count; ++i)
+                indices[i] = rgen.Next(max);
+            Array.Sort(indices, 0, count);
+            for (int i = 0; i < count; ++i)
+                indices[i] += i;
+            Contracts.Assert(Utils.IsIncreasing(0, indices, count, len));
+        }
+
+        /// <summary>
+        /// Copy everything about the <paramref name="src"/>, to <paramref name="dst"/>,
+        /// not just something logically equivalent but possibly with different internal
+        /// structure. This will produce a totally inefficient copy right down to the
+        /// length and contents of the excess indices/values.
+        /// </summary>
+        private static void FullCopy<T>(ref VBuffer<T> src, ref VBuffer<T> dst)
+        {
+            var editor = VBufferEditor.Create(ref dst, src.Length, src.GetValues().Length);
+            var indices = editor.Indices;
+            var values = editor.Values;
+            src.GetIndices().CopyTo(indices);
+            src.GetValues().CopyTo(values);
+            dst = editor.Commit();
+        }
+
+        private static void TestSame(ref VBuffer<float> expected, ref VBuffer<float> actual, Double tol = 0)
+        {
+            TestSame<float>(ref expected, ref actual, FloatEquality((float)tol));
+        }
+
+        private static void TestSame<T>(ref VBuffer<T> expected, ref VBuffer<T> actual, Func<T, T, bool> equalityFunc)
+        {
+            Assert.Equal(expected.Length, actual.Length);
+            Assert.Equal(expected.GetValues().Length, actual.GetValues().Length);
+            Assert.Equal(expected.IsDense, actual.IsDense);
+            if (!expected.IsDense)
+            {
+                for (int i = 0; i < expected.GetIndices().Length; ++i)
+                    Assert.Equal(expected.GetIndices()[i], actual.GetIndices()[i]);
+            }
+            for (int i = 0; i < expected.GetValues().Length; ++i)
+            {
+                var result = equalityFunc(expected.GetValues()[i], actual.GetValues()[i]);
+                if (!result)
+                    Console.WriteLine("Friendly debug here");
+                Assert.True(result, $"Value [{i}] mismatch on expected {expected.GetValues()[i]} vs. actual {actual.GetValues()[i]}");
+            }
+        }
+
+        private static Func<float, float, bool> FloatEquality(float tol)
+        {
+            if (tol == 0)
+                return (i, j) => FloatUtils.GetBits(i) == FloatUtils.GetBits(j);
+            return (i, j) =>
+            {
+                if (FloatUtils.GetBits(i) == FloatUtils.GetBits(j) || Math.Abs(i - j) == 0)
+                    return true;
+                // Seemingly needlessly verbose here for the purpose of setting breakpoints.
+                float comp = Math.Abs(i - j) / (Math.Abs(i) + Math.Abs(j));
+                return comp <= tol;
+            };
+        }
+
+        private void AssertAreEqual(float expected, float actual, int tol)
```

Comment:
We already have this method (with different implementation):
[https://github.com/dotnet/machinelearning/blob/05284d62b167a4a8b98fdfabbba178b7609071b6/test/Microsoft.ML.TestFramework/BaseTestBaseline.cs#L553](https://github.com/dotnet/machinelearning/blob/05284d62b167a4a8b98fdfabbba178b7609071b6/test/Microsoft.ML.TestFramework/BaseTestBaseline.cs#L553)